cdap-master/src/main/java/co/cask/cdap/data/runtime/main/StartupChecker.java

```diff
@@ -0,0 +1,377 @@
+   * directory permissions, wrong hive dependencies, and underlying infrastructure like HDFS being unavailable.
+   *
+   * Other types of errors that may eventually resolve themselves, such as failure to connect to transaction service,
+   * are handled in a separate method.
+   *
+   * This method is a no-op if {@link Constants#STARTUP_CHECKS_ENABLED} is false.
+   *
+   * @return whether the sanity checks succeeded or not
+   */
+  public boolean runSanityChecks() {
+    boolean noErrors = true;
+
+    ConfigurationLogger.logImportantConfig(cConf);
+    if (cConf.getBoolean(Constants.STARTUP_CHECKS_ENABLED)) {
+      noErrors = checkConfig();
+      noErrors = checkHDFS() && noErrors;
+      noErrors = checkYarn() && noErrors;
+      noErrors = checkHBase() && noErrors;
+    }
+
+    return noErrors;
+  }
+
+  // TODO: (CDAP-4517) add more checks, like zookeeper settings.
+  // check cdap config settings. For example, memory.mb settings should be numbers and positive.
+  private boolean checkConfig() {
+    LOG.info("Checking that CDAP config settings are valid.");
+
+    boolean noErrors = true;
+    for (ServiceResourceKeys serviceResourceKeys : systemServicesResourceKeys) {
+      noErrors = noErrors && validatePositiveIntegerAndGet(serviceResourceKeys.memoryKey) != null;
+      noErrors = noErrors && validatePositiveIntegerAndGet(serviceResourceKeys.vcoresKey) != null;
+
+      Integer instances = validatePositiveIntegerAndGet(serviceResourceKeys.instancesKey);
+      Integer maxInstances = validatePositiveIntegerAndGet(serviceResourceKeys.maxInstancesKey);
+      noErrors = noErrors && instances != null && maxInstances != null;
+
+      if (instances != null && maxInstances != null && instances > maxInstances) {
+        LOG.error("{}={} must not be greater than {}={}",
+                  serviceResourceKeys.instancesKey, instances,
+                  serviceResourceKeys.maxInstancesKey, maxInstances);
+        noErrors = false;
+      }
+    }
+
+    if (noErrors) {
+      LOG.info("CDAP configuration check passed.");
+    }
+    return noErrors;
+  }
+
+  // check that the specified key has a positive integer value.
+  // Returns null if it does not and logs an error message. Otherwise it returns the valid value.
+  private Integer validatePositiveIntegerAndGet(String key) {
+    try {
+      int val = cConf.getInt(key);
+      if (val <= 0) {
+        LOG.error("{} has invalid value {}. Must be a positive integer.", key, cConf.get(key));
+        return null;
+      }
+      return val;
+    } catch (Exception e) {
+      LOG.error("{} has invalid value {}. Must be a positive integer.", key, cConf.get(key));
+      return null;
+    }
+  }
+
+  private boolean checkHDFS() {
+    LOG.info("Checking that Hadoop Distributed FileSystem (HDFS) is available.");
+
+    FileSystem hdfs;
+    try {
+      hdfs = FileSystem.get(hConf);
```

Comment:
The reason is, the master may not be using `FileSystem` (CDAP is actually using FileContext, in future it may even be pluggable, but it will always represented by the `LocationFactory` abstraction);